File: android/app/src/main/kotlin/com/network/proxy/vpn/ConnectionManager.kt
package com.network.proxy.vpn

import android.os.Build
import android.util.Log
import com.network.proxy.vpn.socket.CloseableConnection
import com.network.proxy.vpn.socket.Constant
import com.network.proxy.vpn.socket.ProtectSocketHolder.Companion.protect
import com.network.proxy.vpn.util.PacketUtil
import com.network.proxy.vpn.util.ProcessInfoManager
import java.io.IOException
import java.net.InetSocketAddress
import java.net.SocketAddress
import java.nio.ByteBuffer
import java.nio.channels.DatagramChannel
import java.nio.channels.SocketChannel
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ConcurrentMap

/**
 * 管理VPN客户端的连接
 */
class ConnectionManager private constructor() : CloseableConnection {
    //单例
    companion object {
        val instance = ConnectionManager()
    }

    private val table: ConcurrentMap<String, Connection> = ConcurrentHashMap()
    var proxyAddress: InetSocketAddress? = null

    private val DEFAULT_PORTS: List<Int> = listOf(
        80,  // HTTP
        443,  // HTTPS
        8080,  // Common local dev ports
        8000, 8080, 8888, 9000 // Common local dev ports
    )

    override fun closeConnection(connection: Connection) {
        closeConnection(
            connection.protocol, connection.destinationIp, connection.destinationPort,
            connection.sourceIp, connection.sourcePort
        )
    }

    /**
     * 从内存中删除连接，然后关闭套接字。
     *
     */
    fun closeConnection(protocol: Protocol, ip: Int, port: Int, srcIp: Int, srcPort: Int) {
        val key = Connection.getConnectionKey(protocol, ip, port, srcIp, srcPort)
        val connection: Connection? = table.remove(key)
        Log.d(TAG, "close connection $key")

        connection?.let {
            val channel = connection.channel
            try {
                channel?.close()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
    }

    fun getConnection(
        protocol: Protocol, ip: Int, port: Int, srcIp: Int, srcPort: Int
    ): Connection? {
        val key = Connection.getConnectionKey(protocol, ip, port, srcIp, srcPort)
        return getConnectionByKey(key)
    }

    fun getConnectionByKey(key: String?): Connection? {
        return table[key]
    }

    /**
     * 创建tcp连接
     */
    fun createTCPConnection(ip: Int, port: Int, srcIp: Int, srcPort: Int): Connection {
        val key = Connection.getConnectionKey(Protocol.TCP, ip, port, srcIp, srcPort)
        val existingConnection: Connection? = table[key]
        if (existingConnection != null) {
            return existingConnection
        }

        val connection = Connection(Protocol.TCP, srcIp, srcPort, ip, port, this)

        val channel: SocketChannel = SocketChannel.open()
        channel.socket().keepAlive = true
        channel.socket().tcpNoDelay = true
        channel.socket().soTimeout = 0
        channel.socket().receiveBufferSize = Constant.MAX_RECEIVE_BUFFER_SIZE
        channel.configureBlocking(false)

        Log.d(TAG, "created new SocketChannel for $key")

        protect(channel.socket())

        connection.channel = channel

        var socketAddress: SocketAddress? = null
//        if (DEFAULT_PORTS.contains(port)) {
//            socketAddress = proxyAddress
//        }

        connection.isInitConnect = socketAddress != null

        if (socketAddress != null) {
            val connected = channel.connect(socketAddress)
            connection.isConnected = connected
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                //获取进程信息
                ProcessInfoManager.instance.setConnectionOwnerUid(connection)
                Log.d(
                    TAG,
                    "Initiate connecting  " + channel.localAddress + " to remote tcp server: " + channel.remoteAddress
                )
            }
        }

        table[key] = connection
        return connection
    }


    @Throws(IOException::class)
    fun createUDPConnection(ip: Int, port: Int, srcIp: Int, srcPort: Int): Connection {
        val keys = Connection.getConnectionKey(Protocol.UDP, ip, port, srcIp, srcPort)

        val existingConnection: Connection? = table[keys]
        if (existingConnection != null) return existingConnection

        val connection = Connection(Protocol.UDP, srcIp, srcPort, ip, port, this)
        val channel: DatagramChannel = DatagramChannel.open()
        channel.socket().soTimeout = 0
        channel.configureBlocking(false)
        protect(channel.socket())
        connection.channel = channel

        // Initiate connection early to reduce latency
        val ips = PacketUtil.intToIPAddress(ip)
        val socketAddress: SocketAddress = InetSocketAddress(ips, port)
        channel.connect(socketAddress)
        connection.isConnected = channel.isConnected
        table[keys] = connection

        return connection
    }

    /**
     * 添加来自客户端的数据，该数据稍后将在接收到PSH标志时发送到目的服务器。
     */
    fun addClientData(buffer: ByteBuffer, session: Connection): Int {
        return if (buffer.limit() <= buffer.position()) 0 else session.setSendingData(buffer)
    }

    /**
     * 阻止java垃圾收集器收集会话
     */
    fun keepSessionAlive(connection: Connection) {
        val key = Connection.getConnectionKey(
            connection.protocol, connection.destinationIp, connection.destinationPort,
            connection.sourceIp, connection.sourcePort
        )
        table[key] = connection
    }
}

File: ios/ProxyPin/vpn/ConnectionManager.swift
//
//  ConnectionManager.swift
//  ProxyPin
//
//  Created by wanghongen on 2024/9/16.
//

import Foundation
import Network
import os.log

//管理VPN客户端的连接
class ConnectionManager : CloseableConnection{
    //static let instance = ConnectionManager()
    
    private var table: [String: Connection] = [:]
    private let tableQueue = DispatchQueue(label: "ProxyPin.ConnectionManager")

    public var proxyAddress: NWEndpoint?
    
    private let defaultPorts: [UInt16] = [80, 8080, 8888, 443]
    
   
    func getConnection(nwProtocol: NWProtocol, ip: UInt32, port: UInt16, srcIp: UInt32, srcPort: UInt16) -> Connection? {
        let key = Connection.getConnectionKey(nwProtocol: nwProtocol, destIp: ip, destPort: port, sourceIp: srcIp, sourcePort: srcPort)
        return getConnectionByKey(key: key)
    }
    
    func getConnectionByKey(key: String) -> Connection? {
        return tableQueue.sync {
            return table[key]
        }
    }
    
    func createTCPConnection(ip: UInt32, port: UInt16, srcIp: UInt32, srcPort: UInt16) -> Connection {
        let key = Connection.getConnectionKey(nwProtocol: .TCP, destIp: ip, destPort: port, sourceIp: srcIp, sourcePort: srcPort)

        return tableQueue.sync {
            if let existingConnection = table[key] {
                return existingConnection
            }

            let connection = Connection(nwProtocol: .TCP, sourceIp: srcIp, sourcePort: srcPort, destinationIp: ip, destinationPort: port, connectionCloser: self)

            let parameters = NWParameters.tcp
             parameters.allowLocalEndpointReuse = true
             parameters.includePeerToPeer = true

            let endpoint: NWEndpoint
            if defaultPorts.contains(port) {
                endpoint = proxyAddress!
            } else {
                let ipString = PacketUtil.intToIPAddress(ip)
                endpoint = NWEndpoint.hostPort(host: NWEndpoint.Host(ipString), port: NWEndpoint.Port(rawValue: port)!)
            }

            let nwConnection = NWConnection(to: endpoint, using: parameters)

            connection.channel = nwConnection
            connection.isInitConnect = true

            tableQueue.async(flags: .barrier) {
                self.table[key] = connection
            }
            os_log("Created TCP connection %{public}@", log: OSLog.default, type: .default, key)

            return connection
        }

//         nwConnection.stateUpdateHandler = { state in
//             switch state {
//             case .ready:
//                 connection.isConnected = true
//                 os_log("Connected to %{public}@", log: OSLog.default, type: .debug, endpoint.debugDescription)
//             case .failed(let error):
//                 connection.isConnected = false
//                 os_log("Failed to connect: %{public}@", log: OSLog.default, type: .error, error.localizedDescription)
//
//             default:
//                 break
//             }
//         }

    }
    
    func createUDPConnection(ip: UInt32, port: UInt16, srcIp: UInt32, srcPort: UInt16) -> Connection {
        let key = Connection.getConnectionKey(nwProtocol: .UDP, destIp: ip, destPort: port, sourceIp: srcIp, sourcePort: srcPort)

         return tableQueue.sync {
                    if let existingConnection = table[key] {
                        return existingConnection
                    }

               let connection = Connection(nwProtocol: .UDP, sourceIp: srcIp, sourcePort: srcPort, destinationIp: ip, destinationPort: port, connectionCloser: self)
             
            
                let endpoint = NWEndpoint.hostPort(host: NWEndpoint.Host((PacketUtil.intToIPAddress(ip))), port: NWEndpoint.Port(rawValue: port)!)

                let nwConnection = NWConnection(to: endpoint, using: .udp)
                connection.channel = nwConnection

                os_log("Created UDP connection %{public}@", log: OSLog.default, type: .default, key)

                connection.isConnected = true
                tableQueue.async(flags: .barrier) {
                    self.table[key] = connection
               }

                return connection
        }
    }
    
    func closeConnection(connection: Connection) {
        closeConnection(
            nwProtocol: connection.nwProtocol, ip: connection.destinationIp, port: connection.destinationPort,
            srcIp: connection.sourceIp, srcPort: connection.sourcePort
        )
    }
    
    // 从内存中删除连接，然后关闭套接字。
    func closeConnection(nwProtocol: NWProtocol, ip: UInt32, port: UInt16, srcIp: UInt32, srcPort: UInt16) {
        let key = Connection.getConnectionKey(nwProtocol: nwProtocol, destIp: ip, destPort: port, sourceIp: srcIp, sourcePort: srcPort)
        tableQueue.async(flags: .barrier) {
            if let connection = self.table.removeValue(forKey: key) {
                if connection.channel?.state != .cancelled {
                    connection.channel?.cancel()
                    os_log("Closed connection %{public}@", log: OSLog.default, type: .debug, key)
                } else {
                    os_log("Connection %{public}@ is already cancelled", log: OSLog.default, type: .debug, key)
                }
            }
        }
    }
    
    //添加来自客户端的数据，该数据稍后将在接收到PSH标志时发送到目的服务器。
    func addClientData(data: Data, connection: Connection)  {
        guard data.count > 0 else {
            return
        }
        
        connection.addSendData(data: data)
    }
    
    //阻止java垃圾收集器收集会话
    func keepSessionAlive(connection: Connection) {
        let key = Connection.getConnectionKey(
            nwProtocol: connection.nwProtocol,
            destIp: connection.destinationIp,
            destPort: connection.destinationPort,
            sourceIp: connection.sourceIp,
            sourcePort: connection.sourcePort
        )
        tableQueue.async(flags: .barrier) {
            self.table[key] = connection
       }
    }
}
